### DISCUSSION TITLE
Remove respond and log

### DISCUSSION THREAD

**Discussion by mattpocock** [statelyai/xstate#3218] [Ideas]
## Removing `respond`

We should remove `respond` because it's largely useless and doesn't cover most use cases. From an internal discord discussion:

```
I've never used the respond action. It feels strange that you would need it over 'send'.

The pattern I'm imagining is a parent with several invoked children, where each can send an event to the parent and you just want to respond to them.

I can't imagine a practical case where responding immediately to that child would be useful. Maybe doing something first, then responding - but by that time you can't use the respond action.

Or I guess it mimics synchronous read access? Albeit with extra steps.
```

## Removing `log`

We should also remove `log`, which is similarly useless - logging to console is something every JS dev knows how to do, we don't need an extra built-in action to assist them.

## Plan

A plan for this would be to immediately deprecate them in v4 and remove them in v5.

**Comment by Andarist** [comment]
> Or I guess it mimics synchronous read access? Albeit with extra steps.

That is probably one of the reasons why it was added.

I'm OK with removing it since we can't make it type-safe and type-safe alternatives exist (or rather, alternatives to this can be made type-safe)

> We should also remove log, which is similarly useless - logging to console is something every JS dev knows how to do, we don't need an extra built-in action to assist them.

The advantage of `log` is that you can customize the `logger` but... I also don't see why you would actually want to do that. Let's remove it in v5.

**Reply by Steve-OH** [reply]
Regarding `log`, this is related to the discussion here: https://discord.com/channels/795785288994652170/962036015466446918

What would be ideal from my point of view is that rather than having a separate `log` action, logging could be enabled and disabled on a whole-machine basis:

```javascript
const myService = interpret(myMachine, { logger: true }); // turns on logging and sends log output to the console
```
or

```javascript
const myService = interpret(myMachine, { logger: myLogFn}); // turns on logging and sends log output to myLogFn
```

When logging is turned on, _everything_, both events and actions*, is sent to the logger.

*And, if we can get @Andarist to see the light, state value as well.

**Comment by Steve-OH** [comment]
I should add that this isn't strictly about logging, but about "instrumenting" a state machine in general, in the vein of aspect-oriented programming, or the writer monad in functional programming: "Decorate" the machine so that in addition to going about its business, it also does some additional thing at every step, where the definition of that thing is up to the programmer implementing the decoration.

**Comment by Silverwolf90** [comment]
My team has only used `respond` once. Here's a simplified example:

```ts
createMachine({
  initial: 'draft',
  states: {
    draft: {},
    published: {
      id: 'published',
      entry: 'respondPublished',
    },
  },
  on: {
    'publish-requested': '#published',
  },
});
```

**Comment by xylophonehero** [comment]
I'm quite a fan of the log function and especially how we can change the implementation via the logger. It doesn't just need to be used for logging to the console but it can also be used to send to an external service such as Papertrail.

It feels a lot nicer to be able to define logs directly in the machine rather than making a separate actions for each time you want to log a different message.

```
const machine = createMachine({
  entry: 'logMachineStarted',
})

useMachine(myMachine, {
  actions: {
    logMachineStarted: myLogFunction('machine started')
  }
}

// Better

const machine = createMachine({
  entry: log('machineStarted'),
})

useMachine(myMachine, {
  logger: myLogFunction
}
```

**Reply by davidkpiano** [reply]
That's a great point! I'm also in favor now of keeping "log" as a first-class concept.

**Comment by Steve-OH** [comment]
Since this thread has been revived...

Here's an example of what I meant above by "instrumenting" a machine. I have logging actions on all of the various bits and pieces that 
I can get to within each machine, and I can turn logging on/off on a per-machine basis. Here's what the log looks like for the `data-loader` machine, which loads the PWA cache with fresh data on page refresh:

![image](https://user-images.githubusercontent.com/612027/173393048-0c752a10-03b6-4605-bde5-5e4f2be04ca4.png)

I turn the log on and off by setting the log level:

![image](https://user-images.githubusercontent.com/612027/173394483-0843bf06-c8b6-4291-a3a6-8c5f8d232290.png)

This works, but it's tedious to set up, and some things, like actions, are near-impossible to hook into. So some built-in means of instrumenting a machine would simplify things.
